src/fabric.c

```diff
@@ -643,8 +690,32 @@ void fi_ini(void)
 			"Search for providers in specific path (default: "
 			PROVDLDIR ")");
 	fi_param_get_str(NULL, "provider_path", &provdir);
-	if (!provdir)
+	if (!provdir) {
 		provdir = PROVDLDIR;
+		
+		/* Try to find the libraries that are visible by their name
+		 * (LD_LIBRARY_PATH, system paths) */
+		for (i = 0; i < count_of(dl_prov_table); i++) {
+			char *lib;
+
+			if (!dl_prov_table[i])
+				continue;
+
+			if (asprintf(&lib, "%s%s%s", lib_prefix,
+				     dl_prov_table[i], lib_postfix) < 0) {
+				FI_WARN(&core_prov, FI_LOG_CORE,
+				        "asprintf failed to allocate memory\n");
+			}
+
+			ofi_reg_dl_prov(lib);
+
+			free(lib);
+			free(dl_prov_table[i]);
+			dl_prov_table[i] = NULL;
+		}
+
+		goto libdl_done;
+	}
```

Comment:
The above set of changes needs to be adjusted and moved into a separate function.

All known providers are already stored in a provider list.  See ofi_provi, with prov_head / prov_tail tracking the list.  We can iterate over that list and call ofi_reg_dl_prov() on each entry.

We always check for dl providers, regardless of how a specific provider was built.  This allows replacing a built-in provider with a dl provider, to support newer providers with older libraries.  The provider versioning should select the latest version.

Reply:
dl_prov_table contains library names, not provider names (i.e. librxm-fi.so instead of ofi_rxm), so we cannot use prov_head list as is

Reply:
We don't want a duplicate list of the same list of providers.  That's sure to cause maintainability issues.  We need to find a way to use or modify the existing list.

We should look at aligning the utility provider names with the full name, rather than a subset of the name.  That might cause some pain short term, but having the build name differ from the provider name is confusing.

Reply:
Then we should to add some kind of parser to extract short name from provider full name.

Reply:
I think we need to align the names around using the full name always, and avoid the short names.

Reply:
I think the simple solution for now would be to call ofi_has_util_prefix() on the provider name, or check the provider type for OFI_PROV_UTIL, and strip off the "ofi_" from the name when forming the library name.  That avoids changes to the build system.

Reply:
@shefty, do you offer use ordered_prov_names list from function ofi_ordered_provs_init() ?